cns/ipampoolmonitor/ipampoolmonitor.go

```diff
@@ -11,121 +13,193 @@ import (
+	select {
+	case <-ch:
+		return true
+	default:
 	}
 
-	return nil
+	return false
 }
 
-// UpdatePoolLimitsTransacted called by request controller on reconcile to set the batch size limits
-func (pm *CNSIPAMPoolMonitor) UpdatePoolLimitsTransacted(scalarUnits cns.ScalarUnits) {
-	pm.Lock()
-	defer pm.Unlock()
-	pm.scalarUnits = scalarUnits
+func (pm *CNSIPAMPoolMonitor) Start(poolMonitorRefreshMilliseconds int, exitChan <-chan struct{}) error {
+	logger.Printf("[ipam-pool-monitor] Starting CNS IPAM Pool Monitor")
+	for {
+		if stopReconcile(exitChan) {
+			return fmt.Errorf("CNS IPAM Pool Monitor received cancellation signal")
+		}
 
-	// TODO rounding?
-	pm.MinimumFreeIps = int(pm.scalarUnits.BatchSize * (pm.scalarUnits.RequestThresholdPercent / 100))
-	pm.MaximumFreeIps = int(pm.scalarUnits.BatchSize * (pm.scalarUnits.ReleaseThresholdPercent / 100))
+		err := pm.Reconcile()
+		if err != nil {
+			logger.Printf("[ipam-pool-monitor] CRITICAL %v", err)
+		}
 
-	pm.initialized = true
+		time.Sleep(time.Duration(poolMonitorRefreshMilliseconds) * time.Millisecond)
+	}
 }
 
-func (pm *CNSIPAMPoolMonitor) checkForResize(freeIPConfigCount int) int {
-	switch {
-	// pod count is increasing
-	case freeIPConfigCount < pm.MinimumFreeIps:
-		logger.Printf("Number of free IP's (%d) < minimum free IPs (%d), request batch increase\n", freeIPConfigCount, pm.MinimumFreeIps)
-		return increasePoolSize
-
-	// pod count is decreasing
-	case freeIPConfigCount > pm.MaximumFreeIps:
-		logger.Printf("Number of free IP's (%d) > maximum free IPs (%d), request batch decrease\n", freeIPConfigCount, pm.MaximumFreeIps)
-		return decreasePoolSize
+func (pm *CNSIPAMPoolMonitor) Reconcile() error {
+	if pm.initialized {
+		cnsPodIPConfigCount := len(pm.cns.GetPodIPConfigState())
+		allocatedPodIPCount := len(pm.cns.GetAllocatedIPConfigs())
+		pendingReleaseIPCount := len(pm.cns.GetPendingReleaseIPConfigs())
+		availableIPConfigCount := len(pm.cns.GetAvailableIPConfigs()) // TODO: add pending allocation count to real cns
+		freeIPConfigCount := int64(availableIPConfigCount + (int(pm.cachedSpec.RequestedIPCount) - cnsPodIPConfigCount))
+
+		logger.Printf("[ipam-pool-monitor] Checking pool for resize, Pool Size: %v, Goal Size: %v, BatchSize: %v, MinFree: %v, MaxFree:%v, Allocated: %v, Available: %v, Pending Release: %v, Free: %v", cnsPodIPConfigCount, pm.cachedSpec.RequestedIPCount, pm.scalarUnits.BatchSize, pm.MinimumFreeIps, pm.MaximumFreeIps, allocatedPodIPCount, availableIPConfigCount, pendingReleaseIPCount, freeIPConfigCount)
+
+		// if there's a pending change to the spec count, and the pending release state is nonzero,
+		// skip so we don't thrash the UpdateCRD
+		if pm.cachedSpec.RequestedIPCount != int64(len(pm.cns.GetPodIPConfigState())) && len(pm.cns.GetPendingReleaseIPConfigs()) > 0 {
+			return nil
+		}
+
+		switch {
+		// pod count is increasing
+		case freeIPConfigCount < pm.MinimumFreeIps:
+			logger.Printf("[ipam-pool-monitor] Increasing pool size...")
```

Comment:
can some more context be added to logging? like the variables used to trigger the increase (`freeIPConfigCount`, `pm.MinimumFreeIps`). Same for similar cases below